### PR TITLE
Make 'date_trunc' returns the same type as its input

### LINKED ISSUE #6653: `date_trunc` always returns `Timestamp(Nanosecond, None)` which might truncate ranges
### Is your feature request related to a problem or challenge?

Prior to https://github.com/apache/arrow-datafusion/pull/6632 and https://github.com/apache/arrow-datafusion/pull/6638 the `date_trunc` function was inconsistent between reported types and actual types for arrays/scalars

Now it is consistent in the sense that

```
date_trunc(Timestamp(Second, ..)) --> Timestamp(Nanosecond, None)
date_trunc(Timestamp(Millisecond, ..)) --> Timestamp(Nanosecond, None)
date_trunc(Timestamp(Microsecond ..)) --> Timestamp(Nanosecond, None)
date_trunc(Timestamp(Nanosecond ..)) --> Timestamp(Nanosecond, None)
```

However, by using `Timestamp(Nanosecond, None)` everywhere will truncate the range over which `date_trunc` will work -- specifically if there is some value that can be represented using `Timestamp(Second)` that doesn't fit into a `Timestamp(Nanosecond)`  such as a value before `1677-09-21 00:12:43.145224193` or after  `2262-04-11 23:47:16.854775807` date_trunc will error.

### Describe the solution you'd like

We could potentially change the function to keep the same `TimeUnit` of its input:

Namely

```
date_trunc(Timestamp(Second, ..)) --> Timestamp(Second, None)
date_trunc(Timestamp(Millisecond, ..)) --> Timestamp(Millisecond, None)
date_trunc(Timestamp(Microsecond ..)) --> Timestamp(Microsecond, None)
date_trunc(Timestamp(Nanosecond ..)) --> Timestamp(Nanosecond, None)
```

### Describe alternatives you've considered

We could also just leave the current behavior 

### Additional context

See comments https://github.com/apache/arrow-datafusion/pull/6632#issuecomment-1587335750

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #6653

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 1]
```diff
@@ -294,16 +294,24 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
                 }
                 "second" => {
                     // trunc to second
-                    let mill = ScalarValue::TimestampNanosecond(
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:297]: I think we would have to change the values returned by https://github.com/apache/arrow-datafusion/pull/6654/files#diff-67ae8808785b2e651767d7ff67dd7c53be04ca098857b52c82ed19927e071cdaL322 as well as the type returned by `BuiltInScalarFunction::return_type`

otherwise we'll be in the situation again where the scalar implementation does something different than the array implementation

However, I was surprised that there were no tests that failed which reveals a gap in our test coverage. I wrote the test in https://github.com/apache/arrow-datafusion/pull/6655 which does pass on `main` but fails on this PR

**Review by alamb — COMMENTED** [review]
Thanks for jumping on this one @Weijun-H  -- but I think this one needs a little more work

cc @viirya

File: datafusion/physical-expr/src/datetime_expressions.rs (hunk 1 above)
Comment (viirya) [reply datafusion/physical-expr/src/datetime_expressions.rs:297]: And the `signature` of `DateTrunc` should be updated, I think. Currently the second argument is only `Timestamp(Nanosecond, None)`. But I guess that this change wants to make output type as same as input type. So you probably want to add more input types to `signature`.

**Comment by alamb** [comment]
marking this one as a draft -- I think if it is merged to main the tests will fail

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -884,8 +885,13 @@ impl BuiltinScalarFunction {
                 ],
                 self.volatility(),
             ),
-            BuiltinScalarFunction::DateTrunc => Signature::exact(
-                vec![Utf8, Timestamp(Nanosecond, None)],
+            BuiltinScalarFunction::DateTrunc => Signature::one_of(
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:888]: this clause seems unreachable given the change above to match `BuiltinScalarFunction::DateTrunc | BuiltinScalarFunction::DateBin => {`

I would personally suggest keeping this change and removing the `BuiltinScalarFunction::DateTrunc | BuiltinScalarFunction::DateBin => {` change as it make the code more explicy

**Review by alamb — COMMENTED** [review]
Thanks @Weijun-H  ! Looking much closer

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 3]
```diff
@@ -282,54 +341,64 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+            _date_trunc(TimeUnit::Microsecond, v, granularity.as_str(), f, tz_opt)?
+        }
+        ColumnarValue::Scalar(ScalarValue::TimestampMillisecond(v, tz_opt)) => {
+            _date_trunc(TimeUnit::Millisecond, v, granularity.as_str(), f, tz_opt)?
+        }
+        ColumnarValue::Scalar(ScalarValue::TimestampSecond(v, tz_opt)) => {
+            _date_trunc(TimeUnit::Second, v, granularity.as_str(), f, tz_opt)?
+        }
+        ColumnarValue::Array(array) => {
+            let array_type = array.data_type();
```
[the remark this answers is not in the document]
Comment (Weijun-H) [reply datafusion/physical-expr/src/datetime_expressions.rs:356]: I use `f` function pointer, which uses `granularity ` as parameter

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 4]
```diff
@@ -282,54 +341,64 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+        ColumnarValue::Scalar(ScalarValue::TimestampMicrosecond(v, tz_opt)) => {
+            _date_trunc(TimeUnit::Microsecond, v, granularity.as_str(), f, tz_opt)?
+        }
+        ColumnarValue::Scalar(ScalarValue::TimestampMillisecond(v, tz_opt)) => {
+            _date_trunc(TimeUnit::Millisecond, v, granularity.as_str(), f, tz_opt)?
+        }
+        ColumnarValue::Scalar(ScalarValue::TimestampSecond(v, tz_opt)) => {
+            _date_trunc(TimeUnit::Second, v, granularity.as_str(), f, tz_opt)?
+        }
+        ColumnarValue::Array(array) => {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:355]: > For Array, should it return the values based on the input type

It should do the same as is done for the Scalar version.

My understanding is that this PR changes the Scalar version to return a type based on the input types (not TimestampNanoSecond) so the array version should as well

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 5]
```diff
@@ -282,54 +341,64 @@ pub fn date_trunc(args: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+            let array_type = array.data_type();
+            match array_type {
+                DataType::Timestamp(TimeUnit::Second, _) => {
+                    let array = as_timestamp_second_array(array)?;
+                    let array = array
+                        .iter()
+                        .map(|x| {
+                            f(Some(x.unwrap() * 1_000_000_000))
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:363]: I see this PR doesn't introduce the problem, but it appears that `date_trunc` panic's on `unwrap()` 🤔 

```
❯ select date_trunc('hour', '2021-02-01T12:01:02');
+------------------------------------------------------+
| date_trunc(Utf8("hour"),Utf8("2021-02-01T12:01:02")) |
+------------------------------------------------------+
| 2021-02-01T12:00:00                                  |
+------------------------------------------------------+
1 row in set. Query took 0.074 seconds.
❯ select date_trunc('hour', null);
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /Users/alamb/Software/arrow-datafusion/datafusion/physical-expr/src/datetime_expressions.rs:314:35
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I will file a ticket

**Review by alamb — APPROVED** [review]
this looks good to me -- and thought @waitingkuo  or @viirya ?

**Comment by viirya** [comment]
I will take another look today or tomorrow.

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 6]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+                "millisecond" => Some(nano.unwrap() / 1_000_000),
+                "microsecond" => Some(nano.unwrap()),
+                _ => Some(nano.unwrap() / 1_000_000_000),
+            };
+            ScalarValue::TimestampSecond(value, tz_opt.clone())
+        }
+        TimeUnit::Millisecond => {
+            let value = match granularity {
+                "minute" => Some(nano.unwrap() / 1_000_000_000 * 1_000),
+                "second" => Some(nano.unwrap() / 1_000_000_000),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:293]:
```suggestion
                "second" => Some(nano.unwrap() / 1_000_000_000 * 1_000),
```

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 7]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+    let result = match tu {
+        TimeUnit::Second => {
+            let value = match granularity {
+                "minute" => Some(nano.unwrap() / 1_000_000_000),
+                "second" => Some(nano.unwrap() / 1_000_000_000),
+                "millisecond" => Some(nano.unwrap() / 1_000_000),
+                "microsecond" => Some(nano.unwrap()),
+                _ => Some(nano.unwrap() / 1_000_000_000),
+            };
+            ScalarValue::TimestampSecond(value, tz_opt.clone())
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:288]:
```suggestion
            let value = Some(nano.unwrap() / 1_000_000_000);
            ScalarValue::TimestampSecond(value, tz_opt.clone())
```

As output unit is second, smaller granularity can be truncated simply.

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 8]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+        }
+        TimeUnit::Millisecond => {
+            let value = match granularity {
+                "minute" => Some(nano.unwrap() / 1_000_000_000 * 1_000),
+                "second" => Some(nano.unwrap() / 1_000_000_000),
+                "millisecond" => Some(nano.unwrap() / 1_000_000),
+                "microsecond" => Some(nano.unwrap() / 1_000),
+                _ => Some(nano.unwrap() / 1_000_000),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:296]: ditto. For granularity smaller than millisecond can be truncated simply.

```suggestion
                _ => Some(nano.unwrap() / 1_000_000),
```

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 9]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+            };
+            ScalarValue::TimestampMillisecond(value, tz_opt.clone())
+        }
+        TimeUnit::Microsecond => {
+            let value = match granularity {
+                "minute" => Some(nano.unwrap() / 1_000_000_000 * 1_000_000),
+                "second" => Some(nano.unwrap() / 1_000_000 * 1_000),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:303]:
```suggestion
                "second" => Some(nano.unwrap() / 1_000_000_000 * 1_000_000),
```

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 10]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+                "millisecond" => Some(nano.unwrap() / 1_000 * 1_000),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:304]:
```suggestion
                "millisecond" => Some(nano.unwrap() / 1_000_000 * 1_000),
```

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 11]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+                "microsecond" => Some(nano.unwrap() / 1_000),
+                _ => Some(nano.unwrap() / 1_000),
+            };
+            ScalarValue::TimestampMicrosecond(value, tz_opt.clone())
+        }
+        _ => {
+            let value = match granularity {
+                "minute" => Some(nano.unwrap() / 1_000_000_000 * 1_000_000_000),
+                "second" => Some(nano.unwrap() / 1_000_000 * 1_000_000),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:313]:
```suggestion
                "second" => Some(nano.unwrap() / 1_000_000_000 * 1_000_000_000),
```

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 12]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+                "millisecond" => Some(nano.unwrap() / 1_000 * 1_000),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:314]:
```suggestion
                "millisecond" => Some(nano.unwrap() / 1_000_000 * 1_000_000),
```

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 13]
```diff
@@ -262,6 +262,65 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+                "microsecond" => Some(nano.unwrap()),
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:315]:
```suggestion
                "microsecond" => Some(nano.unwrap() / 1_000 * 1_000),
```

**Comment by viirya** [comment]
I'm wondering maybe I misunderstand this function? Because the current change looks not correct mostly to me but all tests passed. 🤔

**Comment by Weijun-H** [comment]
> I'm wondering maybe I misunderstand this function? Because the current change looks not correct mostly to me but all tests passed. 🤔

Seem tests does not cover all cases 🤔

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 14]
```diff
@@ -827,7 +912,7 @@ mod tests {
             (
                 "2020-09-08T13:42:29.190855Z",
                 "second",
-                "2020-09-08T13:42:29.000000Z",
+                "2020-09-08T13:42:29.190855Z",
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:915]: This change doesn't seem correct to me -- the original value was correct (truncated to microsecond). Do you know why it was changed? As written it still has microsecond precision I think

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 15]
```diff
@@ -262,6 +268,53 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+    let scale = match tu {
+        TimeUnit::Second => 1_000_000_000,
+        TimeUnit::Millisecond => 1_000_000,
+        TimeUnit::Microsecond => 1_000,
+        TimeUnit::Nanosecond => 1,
+    };
+
+    if value.is_none() {
+        return Ok(None);
+    }
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:287]: This works fine but I think you can simplify the logic with something like the following (and then avoid all the unwraps)

```suggestion
    let Some(value) = value else {
        return Ok(None);
    }
```

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H  -- this is looking so close

The only remaining question I have is about `date_trunc_test` -- otherwise this looks good to go to me

File: datafusion/physical-expr/src/datetime_expressions.rs (hunk 14 above)
Comment (Weijun-H) [reply datafusion/physical-expr/src/datetime_expressions.rs:917]: https://github.com/apache/arrow-datafusion/blob/b17cf0710938eed1c9f7794d5695bfac89bff580/datafusion/physical-expr/src/datetime_expressions.rs#L217-L223

Previously, the truncation of timestamps to seconds occurred prematurely in `date_trunc_single`, resulting in inaccurate results. Therefore, I have made the necessary adjustments and deferred the logic to the `_date_trunc` function. By implementing this change, the granularity of the timestamps is preserved, and truncation is performed at a later stage to ensure accuracy.

Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:917]: Perhaps then we can update the test to call `_date_trunc`

```
            let result = date_trunc_single(granularity, left).unwrap();
``` 

to 
```
            let result = _date_trunc(granularity, left).unwrap();
```

As written is seems to have the wrong answer

Comment (Weijun-H) [reply datafusion/physical-expr/src/datetime_expressions.rs]: I modified `date_trunc_single`, so it will truncate the timestamp to second when the granularity is second.

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H 

I force pushed this branch because it got caught up in the great "main got force-pushed" debacle (see https://github.com/apache/arrow-datafusion/pull/6775) -- once CI passes I plan to merge it in.

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 16]
```diff
@@ -215,13 +215,17 @@ fn quarter_month(date: &NaiveDateTime) -> u32 {
 }
 
 fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
+    if granularity == "millisecond" || granularity == "microsecond" {
+        return Ok(value);
+    }
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:220]: Hm? Why this change?

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 17]
```diff
@@ -262,6 +266,55 @@ fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
...
+            "millisecond" => Some(nano / 1_000 / 1_000 * 1_000),
+            _ => Some(nano / 1_000),
+        },
+        _ => match granularity {
+            "minute" => Some(nano / 1_000_000_000 / 60 * 1_000_000_000 * 60),
+            "second" => Some(nano / 1_000_000_000 * 1_000_000_000),
+            "millisecond" => Some(nano / 1_000_000 * 1_000_000),
+            "microsecond" => Some(nano / 1_000 * 1_000),
+            _ => Some(nano),
+        },
```
Comment (viirya) [inline datafusion/physical-expr/src/datetime_expressions.rs:313]: Looks correct.

File: datafusion/physical-expr/src/datetime_expressions.rs (hunk 16 above)
Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: I think https://github.com/apache/arrow-datafusion/pull/6654#discussion_r1242712961 is the answer -- perhaps I can remove this to make it easier to understand

Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: I find this very confusing -- I think it is because for those granularities, the conversion to chrono::DateTime and back is unecessary (as only sec/min/hour/etc are non uniform in width)

Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: https://github.com/apache/arrow-datafusion/pull/6783#

Comment (viirya) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: Hmm, is it? https://github.com/apache/arrow-datafusion/pull/6654#discussion_r1242712961 notes about the truncation of timestamps to seconds. But my question is why pulling "millisecond" and "microsecond" earlier which basically avoids `with_nanosecond(0)` only, compared with previous code.

Comment (viirya) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: I think for granularity as "second" | "millisecond" | "microsecond" cases, `with_nanosecond(0)` looks correct in previous code.

It is okay to deferr the logic to the `_date_trunc`, but why only "millisecond" and "microsecond" are treated specially?

Comment (viirya) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: You mean `with_nanosecond` also cleans up millisecond and microsecond in the timestamp?

Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: > You mean with_nanosecond also cleans up millisecond and microsecond in the timestamp? Otherwise nanosecond is smaller granularity than millisecond and microsecond. Why reduction at nanosecond level will affect millisecond and microsecond?

Yes, I think that is correct. 

I played around with it and we can simplify the logic like this which I think may make it easier to understand (the test still pass). If you like this better (I do) I can make it a PR (follow on to https://github.com/apache/arrow-datafusion/pull/6783)

```diff
diff --git a/datafusion/physical-expr/src/datetime_expressions.rs b/datafusion/physical-expr/src/datetime_expressions.rs
index 3a36e8a489..89ec508024 100644
--- a/datafusion/physical-expr/src/datetime_expressions.rs
+++ b/datafusion/physical-expr/src/datetime_expressions.rs
@@ -215,42 +215,49 @@ fn quarter_month(date: &NaiveDateTime) -> u32 {
 }
 
 fn date_trunc_single(granularity: &str, value: i64) -> Result<i64> {
-    if granularity == "millisecond" || granularity == "microsecond" {
-        return Ok(value);
-    }
+    let value = timestamp_ns_to_datetime(value).ok_or_else(|| {
+        DataFusionError::Execution(format!("Timestamp {value} out of range"))
+    })?;
+
+    let value = Some(value);
 
-    let value = timestamp_ns_to_datetime(value)
-        .ok_or_else(|| {
-            DataFusionError::Execution(format!("Timestamp {value} out of range"))
-        })?
-        .with_nanosecond(0);
     let value = match granularity {
-        "second" => value,
-        "minute" => value.and_then(|d| d.with_second(0)),
+        "millisecond" => value,
+        "microsecond" => value,
+        "second" => value.and_then(|d| d.with_nanosecond(0)),
+        "minute" => value
+            .and_then(|d| d.with_nanosecond(0))
+            .and_then(|d| d.with_second(0)),
         "hour" => value
+            .and_then(|d| d.with_nanosecond(0))
             .and_then(|d| d.with_second(0))
             .and_then(|d| d.with_minute(0)),
         "day" => value
+            .and_then(|d| d.with_nanosecond(0))
             .and_then(|d| d.with_second(0))
             .and_then(|d| d.with_minute(0))
             .and_then(|d| d.with_hour(0)),
         "week" => value
+            .and_then(|d| d.with_nanosecond(0))
             .and_then(|d| d.with_second(0))
             .and_then(|d| d.with_minute(0))
             .and_then(|d| d.with_hour(0))
             .map(|d| d - Duration::seconds(60 * 60 * 24 * d.weekday() as i64)),
         "month" => value
+            .and_then(|d| d.with_nanosecond(0))
             .and_then(|d| d.with_second(0))
             .and_then(|d| d.with_minute(0))
             .and_then(|d| d.with_hour(0))
             .and_then(|d| d.with_day0(0)),
         "quarter" => value
+            .and_then(|d| d.with_nanosecond(0))
             .and_then(|d| d.with_second(0))
             .and_then(|d| d.with_minute(0))
             .and_then(|d| d.with_hour(0))
             .and_then(|d| d.with_day0(0))
             .and_then(|d| d.with_month(quarter_month(&d))),
         "year" => value
+            .and_then(|d| d.with_nanosecond(0))
             .and_then(|d| d.with_second(0))
             .and_then(|d| d.with_minute(0))
             .and_then(|d| d.with_hour(0))

```

Comment (viirya) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: Yea, I think it looks clearer.

Comment (alamb) [reply datafusion/physical-expr/src/datetime_expressions.rs:220]: https://github.com/apache/arrow-datafusion/pull/6789
